src/zenml/orchestrators/publish_utils.py

```diff
@@ -92,25 +119,78 @@ def publish_failed_pipeline_run(
     )
 
 
+def publish_pipeline_run_status_update(
+    pipeline_run_id: "UUID",
+    status: ExecutionStatus,
+    end_time: Optional[datetime] = None,
+) -> "PipelineRunResponse":
+    """Publishes a pipeline run status update.
+
+    Args:
+        pipeline_run_id: The ID of the pipeline run to update.
+        status: The new status for the pipeline run.
+        end_time: The end time for the pipeline run. If None, will be set to current time
+            for finished statuses.
+
+    Returns:
+        The updated pipeline run.
+    """
+    if end_time is None and status.is_finished:
+        end_time = utc_now()
+
+    return Client().zen_store.update_run(
+        run_id=pipeline_run_id,
+        run_update=PipelineRunUpdate(
+            status=status,
+            end_time=end_time,
+        ),
+    )
+
+
 def get_pipeline_run_status(
-    step_statuses: List[ExecutionStatus], num_steps: int
+    run_status: ExecutionStatus,
+    step_statuses: List[ExecutionStatus],
+    num_steps: int,
 ) -> ExecutionStatus:
     """Gets the pipeline run status for the given step statuses.
 
     Args:
+        run_status: The status of the run.
         step_statuses: The status of steps in this run.
         num_steps: The total amount of steps in this run.
 
     Returns:
         The run status.
     """
-    if ExecutionStatus.FAILED in step_statuses:
-        return ExecutionStatus.FAILED
+    # Any finished state
+    if run_status.is_finished:
+        return run_status
+
+    # STOPPING state
+    if run_status == ExecutionStatus.STOPPING:
+        if all(status.is_finished for status in step_statuses):
+            return ExecutionStatus.STOPPED
+        else:
+            return ExecutionStatus.STOPPING
+
+    # RUNNING and INITIALIZING states
     if (
-        ExecutionStatus.RUNNING in step_statuses
-        or len(step_statuses) < num_steps
+        run_status == ExecutionStatus.RUNNING
+        or run_status == ExecutionStatus.INITIALIZING
     ):
-        return ExecutionStatus.RUNNING
+        if ExecutionStatus.FAILED in step_statuses:
+            return ExecutionStatus.FAILED
+        elif ExecutionStatus.STOPPED in step_statuses:
+            if all(status.is_finished for status in step_statuses):
+                return ExecutionStatus.STOPPED
```

Comment:
How can a step be in status `STOPPED` with a run not in status `STOPPING/STOPPED`?